Secure Expense Dashboard with Express and Okta Blog (#1584)
oktadeveloper/okta-blog: _source/_posts/2025-07-08-express-authentication.md

---
layout: blog_post

> title: "Secure Your Express App with Okta and OAuth 2.0 Authentication"

Reviewer: The title in the repo Readme doesn't match this post.

Is Okta a differentiator somehow for this post? There is usually not a need to add Okta unless we are featuring a new productline or something. this is posted in OktaDev blog and GH org, Okta is implied already

by: advocate
communities: [javascript]
description: "Build a secure Express app with Okta using OIDC and OAuth 2.0 PKCE"
tags: [express, node, passport, oidc, oauth, pkce, javascript, authentication]
image: blog/express-okta-authentication/express-okta-authentication-social-image.jpeg
type: conversion
github: https://github.com/oktadev/okta-express-oauth-example
---

Every web application needs authentication, but building it yourself is risky and time-consuming. Instead of starting from scratch, you can integrate Okta to manage user identity and pair Passport with the openid-client library in Express to simplify and secure the login flow. In this tutorial, you'll build a secure, role-based expense dashboard where users can view their expenses tailored to their team.

Check out the complete source code on [GitHub](https://github.com/oktadev/okta-express-oauth-example) and get started without setting it up from scratch. 

**Table of Contents**{: .hide }
* Table of Contents
{:toc}

## Why use Okta for authentication 

Building an authentication system and handling credentials, sessions, and tokens is highly insecure and exposes your application to serious vulnerabilities.

> Okta provides a secure, scalable, and standards-based solution using OpenID Connect (OIDC) and OAuth 2.0. It also integrates seamlessly with Express and Passport, and allows you to fetch tokens.

Reviewer: ```suggestion
Okta provides a secure, scalable, and standards-based solution using OpenID Connect (OIDC) and OAuth 2.0. It also integrates seamlessly with OIDC client libraries for your favorite tech stack and allows you to fetch tokens.
```

Reviewer: let's deprioritize passport because it is deprioritized

> ### Why use PKCE in OAuth 2.0

Reviewer: ```suggestion
### Why use PKCE in OAuth 2.0

```

You'll also use OpenID Connect (OIDC) through openid-client library for authentication. Then, you'll map each user's email from the ID token to a team. The dashboard applies principles of least privilege and displays expenses by team, so each user sees only their department's spending.

**Prerequisites**

* Node.js installed (v22+ recommended)

* [Okta Integrator Free Plan org](https://developer.okta.com/signup/)

## Create your Express project and install dependencies

Create a new project folder named `express-project-okta`, and open a terminal window in the project folder.

Initialize a new Node.js project:

```sh
npm init -y
```

Install the required packages:

```sh
npm install express@5.1 passport@0.7 openid-client@6.6 express-session@1.18 ejs@3.1 express-ejs-layouts@2.5 dotenv
```

Now, install the development dependencies:

```sh
npm install --save-dev nodemon
``` 

**What do these dependencies do?** 

These installed packages become your Express project's dependencies.

* **`express`**: Handles routing and HTTP middleware for your web app

* **`passport`**: Sets up and maintains server-side sessions

> * **`openid-client`**: A server-side OpenID Relying Party implementation for Node.js runtime, including PKCE support

Reviewer: ```suggestion
* **`openid-client`**: A Node.js OIDC library that supports PKCE we'll use to handle the OAuth handshake and token exchange
```

      amount: 1000,
    },
  ],
  marketing: [
    {
      name: "Alice Johnson",
      item: "Payroll Processing",
      amount: 750,
    },
    {
      name: "Carol Lee",
      item: "Compliance Training",
      amount: 400,
    },
    {
      name: "Eve Martinez",
      item: "Team Lunch",
      amount: 200,
    },
    {
      name: "Frank Wilson",
      item: "Team Offsite",
      amount: 850,
    },
  ],
  legal: [
    {
      name: "Grace Li",
      item: "Event Booth",
      amount: 1100,
    },
    {
      name: "David Kim",
      item: "Product Launch Campaign",
      amount: 1300,
    },
    {
      name: "Bob Smith",
      item: "Conference Travel",
      amount: 1250,
    },
    {
      name: "Henry Zhang",
      item: "Team Lunch",
      amount: 170,
    },
  ],
  "dev-advocacy": [
    {
      name: "Eve Martinez",
      item: "Internet",
      amount: 280,
    },
    {
      name: "Frank Wilson",
      item: "Payroll Processing",
      amount: 720,
    },
    {
      name: "Grace Li",
      item: "Compliance Training",
      amount: 500,
    },
    {
      name: "Alice Johnson",
      item: "Team Offsite",
      amount: 950,
    },
  ],
};
```

> ### Create an auth file to handle authentication

Reviewer: ```suggestion
### Create a file to handle authentication
```

The authentication logic is kept in a separate file, `auth.js`, to keep your server organized and maintainable. This file manages OpenID Connect authentication with [openid-client](https://www.passportjs.org/packages/openid-client), including PKCE support. It sets up the OIDC client, handles login and logout, processes callbacks, and provides middleware to protect routes.

```javascript
import * as client from "openid-client";
import "dotenv/config";

import { userTeamMap, ALL_TEAMS_NAME } from './utils.js';

> export function ensureAuthenticated(req, res, next) {

Reviewer: We need to discuss what's going on here. This should be the main focus of the blog. What is the `ensureAuthticated` function, how does PKCE work in this library, what are you doing with the callbackURLParams function, etc etc

  if (req.isAuthenticated()) {
    return next();
  }
  res.redirect("/login");
}

function getCallbackUrlWithParams(req) {
  const host = req.headers["x-forwarded-host"] || req.headers.host || "localhost";
  const protocol = req.headers["x-forwarded-proto"] || req.protocol;
  const currentUrl = new URL(`${protocol}://${host}${req.originalUrl}`);
  return currentUrl;
}

> function getModifiedTeam(team) {

Reviewer: We should move this out somewhere else and leave this to be purely about authn

      expectedState: pkce.state,
    });

    const { name, email } = tokenSet.claims();
    const teams = getModifiedTeam(userTeamMap[email]);

    const userProfile = {
      name,
      email,
      teams,
      idToken: tokenSet.id_token,
    };

    delete req.session.pkce;

    req.logIn(userProfile, (err) => {
      if (err) {
        return next(err);
      }

      return res.redirect("/dashboard");
    });
  } catch (error) {
    console.error("Authentication error:", error.message);
    return res.status(500).send(`Authentication failed: ${error.message}`);
  }
}

export async function logout(req, res) {
  try {
    const openIdClientConfig = await getClientConfig();

    const id_token_hint = req.user?.idToken;

    const logoutUrl = client.buildEndSessionUrl(openIdClientConfig, {
      id_token_hint,
      post_logout_redirect_uri: process.env.POST_LOGOUT_URL,
    });

    req.logout((err) => {
      if (err) return next(err);

      req.session.destroy((err) => {
        if (err) return next(err);
        res.redirect(logoutUrl);
      });
    });
  } catch (error) {
    res.status(500).send('Something went wrong during logout.');
  }
}
```

> ### Set up the express routes file 

Reviewer: ```suggestion
### Set up routing in Express
```

import { dummyExpenseData } from './utils.js';

const router = express.Router();

router.get("/", (req, res) => {
  res.render("home", { title: "Home", user: req.user });
});

router.get("/login", login);

router.get("/authorization-code/callback", authCallback);

router.get("/profile", ensureAuthenticated, (req, res) => {
  res.render("profile", { title: "Profile", user: req.user });
});

router.get("/dashboard", ensureAuthenticated, (req, res) => {
  const team = req.user?.teams || [];

  res.render("dashboard", {
    title: "Dashboard",
    user: req.user,
    team,
  });
});

router.get("/team/:id", ensureAuthenticated, (req, res) => {
  const teamId = req.params.id;
  const teamList = req.user?.teams || [];

  const team = teamList.find((team) => team.id === teamId);
  if (!team) {
    return res.status(404).send("Team not found");
  }

  const expenses = dummyExpenseData[teamId];
  const total = expenses.reduce((sum, exp) => sum + exp.amount, 0);

  res.render("expenses", {
    title: team.name,
    user: req.user,
    team,
    expenses,
    total,
  });
});

router.get("/logout", logout);

export default router;
```

> ### Add views in the app

Reviewer: ```suggestion
### Add EJS views in Express
```

<h3>Your Teams</h3>
<% if (team && team.length > 0) { %>
<ul class="list-group">
  <% teams.forEach(team => { %>
  <li class="list-group-item d-flex justify-content-between align-items-center">
    <%= team.label %>
    <a href="/teams/<%= team.id %>" class="btn btn-primary btn-sm">View</a>
  </li>
  <% }) %>
</ul>
<% } else { %>
<p>You are not part of any teams yet.</p>
<% } %>
```

**expenses.ejs**
  
The EJS template renders the team info and expenses data in a tabular format.

```javascript
<h1><%= team.label %></h1>
<div>Welcome to the <p class="team-heading"><%= team.label %></p> team page.</div>
<br/>
<% if (expenses && expenses.length > 0) { %>
<h3>Expenses</h3>
<table class="table table-bordered">
  <thead>
    <tr>
      <th>Name</th>
      <th>Item</th>
      <th>Amount ($)</th>
    </tr>
  </thead>
  <tbody>
      <% expenses.forEach(exp => { %>
    <tr>
      <td><%= exp.name %></td>
      <td><%= exp.item %></td>
      <td><%= exp.amount %></td>
    </tr>
    <% }) %>
  </tbody>
</table>
<div class="alert alert-info"><h6 style="display: inline-block; margin: 0;">Total:</h6> $<%= total %></div>
<% } else { %>
<p>No expenses found for this team.</p>
<% } %>
```

> ## Run the app

Reviewer: ```suggestion
## Run the Express app with authentication
```

---
layout: blog_post
title: "Secure Your Express App with Okta and OAuth 2.0 Authentication"
author: akanksha-bhasin 
by: advocate
communities: [javascript]

> description: "Build a secure expense dashboard using Express, Passport, and Okta using OIDC and OAuth 2.0 PKCE"

Reviewer: ```suggestion
description: "Build a secure Express app with Passport and Okta using OIDC and OAuth 2.0 + PKCE"
```

type: conversion
github: https://github.com/oktadev/okta-express-expense-dashboard-example
---

Every web application needs authentication, but building it yourself is risky and time-consuming. Instead of starting from scratch, you can integrate Okta to manage user identity and use Passport with Express to simplify and secure the login flow. In this tutorial, you'll build a secure, passwordless, role-based expense dashboard where users can view their expenses tailored to their team.

Check out the complete source code on [GitHub](https://github.com/oktadev/okta-express-expense-dashboard-example) and get started without setting it up from scratch. 

**Table of Contents**{: .hide }
* Table of Contents
{:toc}

## Why use Okta for authentication 

Building an authentication system and handling credentials, sessions, and tokens is highly insecure and exposes your application to serious vulnerabilities.

Okta provides a secure, scalable, and standards-based solution using OpenID Connect (OIDC) and OAuth 2.0. It also integrates seamlessly with Express and Passport, supports passwordless login using FIDO2 and WebAuthn, and allows you to customise ID tokens with team-specific claims.

> ### Why Use PKCE in OAuth 2.0

Reviewer: ```suggestion
### Why use PKCE in OAuth 2.0
```

Sentence casing in headers per style

> To further strengthen security, this project uses PKCE (Proof Key for Code Exchange), defined in [RFC 7636](https://www.rfc-editor.org/rfc/rfc7636). PKCE is a security extension to the Authorization Code flow. Developers initially designed PKCE for mobile apps, but experts now recommend it for all OAuth clients, including web apps. It helps prevent CSRF and authorization code injection attacks and makes it useful for every type of OAuth client, even web apps that use client authentication. As OAuth 2.0 has steadily evolved, security best practices have also advanced. [RFC 9700: Best Current Practice for OAuth 2.0 Security](https://www.rfc-editor.org/rfc/rfc9700.html) captures the consensus on the most effective and secure implementation strategies. Additionally, the upcoming OAuth 2.1 draft requires PKCE for all authorization code flows, reinforcing it as a baseline security standard.

Reviewer: ```suggestion
To further strengthen security, this project uses PKCE (Proof Key for Code Exchange), defined in [RFC 7636](https://www.rfc-editor.org/rfc/rfc7636). PKCE is a security extension to the Authorization Code flow. Developers initially designed PKCE for mobile apps, but experts now recommend it for all OAuth clients, including web apps. It helps prevent CSRF and authorization code injection attacks and makes it useful for every type of OAuth client, even confidential clients such as web apps that use client secrets. As OAuth 2.0 has steadily evolved, security best practices have also advanced. [RFC 9700: Best Current Practice for OAuth 2.0 Security](https://www.rfc-editor.org/rfc/rfc9700.html) captures the consensus on the most effective and secure implementation strategies. Additionally, the upcoming OAuth 2.1 draft requires PKCE for all authorization code flows, reinforcing it as a baseline security standard.
```

With Okta, you can implement modern authentication features and focus on your application logic without worrying about authentication infrastructure.

> ##  Team-based expense dashboard using Express, Passport, and OAuth 2.0

Reviewer: ```suggestion
##  A secure app using Express, Passport, and OAuth 2.0
```

> For any growing organization, tracking expenses by team isn't just helpful, it's essential. Let's build an expense dashboard where users log in with Okta and view spending data. Whether in Finance, Marketing, or Support, each team sees a scoped view of expenses tied to their role. The users get clear visibility into what's being spent, by whom, and why.

Reviewer: ```suggestion
Let's build an expense dashboard where users log in with Okta and view spending data based on their role. Whether in Finance, Marketing, or Support, each team sees a scoped view of expenses using Role-based Access Control (RBAC). To keep things minimal in this demo project, we'll define roles and users directly in the app.
```

> Managers get transparency into their team's spending. At the same time, admins maintain a secure, centralized overview across all teams to track budgets and spending patterns and ensure accountability, all without the headache of building authentication and security from scratch.

Reviewer: ```suggestion
```

We don't need to explain why managers use an expense dashboard. realistically enterprises use ERPs that handle this (they wouldn't build their own), or in new B2B app scenarios, people understand the need for accounting

> You'll also use OpenID Connect (OIDC) through Passport and the openid-client library. Then, you'll map each user's email from the ID token to a team. The dashboard displays expenses by team, so each user sees only their department's spending.

Reviewer: ```suggestion
You'll also use OpenID Connect (OIDC) through Passport and the openid-client library. Then, you'll map each user's email from the ID token to a team. The dashboard applies principles of least privilege and displays expenses by team, so each user sees only their department's spending.
```

```sh
npm install --save-dev nodemon
``` 

**What do these dependencies do?** 

These installed packages become your Express project's dependencies.

* **`express`**: Handles routing and HTTP middleware for your Node.js web app

* **`passport`**: Provides a flexible authentication framework

* **`openid-client`**: A server-side OpenID Relying Party implementation for Node.js runtime, including PKCE support

* **`express-session`**: Manages user sessions on the server

* **`dotenv`**: Loads environment variables from a `.env` file

* **`ejs`**: Enables dynamic HTML rendering using embedded JavaScript templates

* **`express-ejs-layouts`**: Adds layout support to EJS, helping manage common layout structures across views

## Configure environment variables for OIDC authentication

Create a `.env` file in the root directory with placeholders for your Okta configuration.

```
OKTA_ISSUER=https://{yourOktaDomain}/oauth2/default 
OKTA_CLIENT_ID={yourClientId}
OKTA_CLIENT_SECRET={clientSecret}
APP_BASE_URL=http://localhost:3000
POST_LOGOUT_URL=http://localhost:3000
```

In the next step, you'll get these values from your Okta Admin Console.

## Create the Okta OIDC application

> 1. Sign up for a free [Integrator Free Plan](https://developer.okta.com/signup/). If you already have an account, [login](https://developer.okta.com/login/) to the [Okta Developer Console](https://developer.okta.com/signup/). 

Reviewer: Please replace this using the embed as discussed in our team Slack channel

2. In the Okta Admin Console, navigate to **Applications** \> **Create App Integration**.

3. Choose:
   * **Sign-in method:** OIDC \- OpenID Connect

   * **Application type:** Web Application

4. Click **Next**.  
                             
5. Fill in:
   * **App integration name:** (e.g., `My Web App`)

   * **Sign-in redirect URIs:** `http://localhost:3000/authorization-code/callback`

   * **Sign-out redirect URIs:** `http://localhost:3000`

   * **Assignments:** Allow everyone in your organization to access.

6. After creating the app, click the edit button under Client Credentials and enable **Require PKCE as additional verification**.

7. Copy the **Client ID**, **Client Secret**, and **Okta Domain** and add them to your `.env` file.

> ## Set up passwordless login using FIDO2 and WebAuthn with Okta 

Reviewer: Why are we doing this? I don't see the value this section brings to the post. There are definitely times when discussing passwordless and adding it into a post makes sense. In this case, it appears slapped on. As previously stated, this is not something ISVs configure for their app submissions.

As we discussed already, please remove this from the post and the GitHub project readme.

Add the **FIDO2 (WebAuthn)** authenticator in Okta to enable passwordless login and follow the [Okta documentation](https://help.okta.com/oie/en-us/content/topics/identity-engine/authenticators/configure-webauthn.htm) for complete enrollment and policy configuration instructions.

>  **Note:** When logging in for the first time, make sure you, as the admin, and all assigned users enroll in FIDO authentication. To simplify this process, you can create a user group containing all team members and assign it to the Web App. Then include the group in your enrollment policy, which you can customize as needed.
 

## Building the Express app 

> Create an `index.js` file in your project root. It serves as the main entry point for your application. Use it to initialize the Express app, configure session handling, set up [Passport](https://www.passportjs.org/) to use OIDC, and set up the routes. 

Reviewer: ❓

import session from 'express-session';
import passport from 'passport'; 
import routes from './routes.js';
import expressLayouts from 'express-ejs-layouts';

const app = express();

app.set('view engine', 'ejs');
app.use(expressLayouts); 
app.set('layout', 'layout');
app.use(express.urlencoded({ extended: false }));

app.use(session({
  secret: "your-hardcoded-secret",
  resave: false,
  saveUninitialized: true,
}));

app.use(passport.initialize());
app.use(passport.session());

passport.serializeUser(function (user, done) {
  done(null, user);
});

passport.deserializeUser(function (obj, done) {
  done(null, obj);
});

app.use('/', routes);

app.listen(3000, () => {
  console.log('Server listening on http://localhost:3000');
});
```

### Define team mapping and sample expenses 

Create a `utils.js` file to serve as a data module for your project. This file includes a user-to-team mapping and has dummy expense data for each team, covering all teams configured for testing in your web app.

The application determines the user's team context from the email claim in the ID token and filters the expense list accordingly, so the dashboard displays only that team's data.

> To customize the data, open utils.js and update the following objects:

Reviewer: Is this rendering how you want? for example, should it be bullet points? Should the variables use inline code markdown?

Author: Thanks for the suggestion, have updated it.

ALL_TEAMS_NAME - an array listing all teams in your organization

userTeamMap - maps each user's email (or "admin" for full access) to a specific team

dummyExpenseData - contains sample expense data for each team

```javascript
export const ALL_TEAMS_NAME = ["finance", "hr", "legal", "marketing", "dev advocacy"];

> export const userTeamMap = {

Reviewer: Very very unfortunately, we cannot use these for license reasons.

I recommend using some fake names and I'd also change the domain to something totally fake.

I bought the domain `task-vantage.com` you can use that (or use a invalid TLD like company.fake)

Author: Sure, thanks Alisa. I will change it.
